Preparation for Topology migration to GA for CSI migration

Jiawei0227 opened the pull request:
This also includes a change on CSI migration TranslateCSIToInTree
where we remove the CSI topology and add Kubernetes Topology to
the NodeAffinity


**What type of PR is this?**
/kind bug
/kind cleanup


**What this PR does / why we need it**:
This PR is a subset work from https://github.com/kubernetes/kubernetes/pull/97746 with some comments addressed and test.
Specifically, this PR prepares for the migration from `failure-domain.beta.kubernetes.io/zone` topology to GA version which is `topology.kubernetes.io/zone`.

Besides, this PR also contains a behavior change in the csi-translation library TranslateCSIToInTree for gce-pd where we remove the CSI topology and inserts the kubernetes topology. This surfaces two purpose: 1. Fix a potential rollback issue. 2. Prepare for the fully deprecation of PV label admission-controller.

**Which issue(s) this PR fixes**:
<!--
*Automatically closes linked issue when PR is merged.
Usage: `Fixes #<issue number>`, or `Fixes (paste link of issue)`.
_If PR is about `failing-tests or flakes`, please post the related issues/tests in a comment and do not use `Fixes`_*
-->
Fixes #

Partially fix https://github.com/kubernetes/kubernetes/issues/92237

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:
<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below. If the PR requires additional action from users switching to the new release, include the string "action required".

For more information on release notes see: https://git.k8s.io/community/contributors/guide/release-notes.md
-->
```release-note
Remove CSI topology from migrated in-tree gcepd volume.
```

**Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.**:

<!--
This section can be blank if this pull request does not require a release note.

When adding links which point to resources within git repositories, like
KEPs or supporting documentation, please reference a specific commit and avoid
linking directly to the master branch. This ensures that links reference a
specific point in time, rather than a document that may change over time.

See here for guidance on getting permanent links to files: https://help.github.com/en/articles/getting-permanent-links-to-files

Please use the following format for linking documentation:
- [KEP]: <link>
- [Usage]: <link>
- [Other doc]: <link>
-->
```docs

```

/sig storage
/assign @msau42
/cc @wongma7 @mattcary @verult  

The follow-up for this PR includes:
1. Upgrade csi-translation-lib to the newer version on csi-provisioner
2. Upgrade in-tree volume plugins to use the GA label

Jiawei0227 committed:
Prepare for Topology migration to GA from CSI migration

This also includes a change on CSI migration TranslateCSIToInTree
where we remove the CSI topology and add Kubernetes Topology to
the NodeAffinity

Jiawei0227 commented on staging/src/k8s.io/cloud-provider/volume/helpers/zones.go:
```diff
@@ -135,7 +138,7 @@ func SelectZonesForVolume(zoneParameterPresent, zonesParameterPresent bool, zone
 
 	if (len(allowedTopologies) > 0) && (allowedZones.Len() == 0) {
-		return nil, fmt.Errorf("no matchLabelExpressions with %s key found in allowedTopologies. Please specify matchLabelExpressions with %s key", v1.LabelFailureDomainBetaZone, v1.LabelFailureDomainBetaZone)
+		return nil, fmt.Errorf("no matchLabelExpressions with %s key found in allowedTopologies. Please specify matchLabelExpressions with %s key", v1.LabelTopologyZone, v1.LabelTopologyZone)
```
> Is this logic shared by other clouds too? Will this cause other clouds' Provision() to use GA labels?

This function `SelectZoneForVolume` is shared by all in-tree plugins and is being called in in-tree code. This function only picks a zone value from the labels. So it does not matter if the value comes from GA or beta label. For example:

If a node has `failure-domain.beta.kubernetes.io/zone=us-west-2` and `topology.kubernetes.io/zone=us-west-2`. This function will just return the `us-west-2` so it should not matter here.

Jiawei0227 commented on staging/src/k8s.io/csi-translation-lib/translate_test.go:
```diff
@@ -28,10 +28,14 @@ import (
 var (
-	defaultZoneLabels = map[string]string{
+	kubernetesBetaTopologyLabels = map[string]string{
 		v1.LabelFailureDomainBetaZone:   "us-east-1a",
```
> we should keep test cases with beta labels too

Done. I also add new test to test GA label

msau42 commented on plugin/pkg/admission/storage/persistentvolume/label/admission.go:
```diff
@@ -172,15 +172,31 @@ func (l *persistentVolumeLabel) findVolumeLabels(volume *api.PersistentVolume) (
+	domain, domainOK := existingLabels[v1.LabelTopologyZone]
+	region, regionOK := existingLabels[v1.LabelTopologyRegion]
+	// If they dont have GA labels we should check for failuredomain beta labels
+	// TODO: remove this once all the cloud provider change to topology labels
```
nit: "ga labels" may avoid some confusion

Jiawei0227 replied:
Done

Jiawei0227 commented on staging/src/k8s.io/csi-translation-lib/plugins/in_tree_volume.go:
```diff
@@ -124,38 +124,92 @@ func addTopology(pv *v1.PersistentVolume, topologyKey string, zones []string) er
-// translateTopology converts existing zone labels or in-tree topology to CSI topology.
+// removeTopology removes the topology from the given PV. Return false
+// if the topology key is not found
+func removeTopology(pv *v1.PersistentVolume, topologyKey string) bool {
```
> do we have unit tests for this?

Added. I also add tests for translateTopologyFromCSIToInTree which should also test this function from an e2e perspective.

Jiawei0227 reviewed (commented):
Hey @msau42 , thanks for your review! Sorry I have to close the old one and open this new one.

I think to make it cleaner, I am focusing this PR on the csi-translation-lib fix first and I also include some preparation work for topology label GA promotion. I will work on the actual label migration for gcepd as soon as this PR merges.

I have copied your comment and replied below. Please take a look. Thanks!

Jiawei0227 commented on staging/src/k8s.io/csi-translation-lib/plugins/gce_pd.go:
```diff
@@ -286,7 +291,10 @@ func (g *gcePersistentDiskCSITranslator) TranslateCSIPVToInTree(pv *v1.Persisten
+	if err := translateTopologyFromCSIToInTree(pv, GCEPDTopologyKey); err != nil {
+		return nil, fmt.Errorf("failed to translate topology: %v", err)
+	}
 
```
@wongma7 FYI.

msau42 commented on staging/src/k8s.io/cloud-provider/volume/helpers/zones_test.go:
```diff
@@ -898,7 +898,7 @@ func TestSelectZonesForVolume(t *testing.T) {
 		},
 
-		// Key specified in AllowedTopologies is not LabelFailureDomainBetaZone [Fail]
+		// Key specified in AllowedTopologies is not LabelTopologyZone [Fail]
```
Can you keep one test case with AllowedTopologies using the beta label?

Jiawei0227 replied:
Done. Add two test cases one for node using beta label one for AllowedTopologies using beta label

msau42 commented on staging/src/k8s.io/csi-translation-lib/plugins/in_tree_volume.go:
```diff
@@ -165,6 +219,57 @@ func translateTopology(pv *v1.PersistentVolume, topologyKey string) error {
+	removeTopology(pv, topologyKey)
+
+	// Check if Kubernetes Topology already exist
+	topologyZoneValue := getTopologyValues(pv, v1.LabelTopologyZone)
```
would it be worth it to have a wrapper function that returns either ga or beta topology key/value?

Jiawei0227 replied:
Done

msau42 commented on staging/src/k8s.io/csi-translation-lib/plugins/gce_pd.go:
```diff
@@ -286,7 +291,10 @@ func (g *gcePersistentDiskCSITranslator) TranslateCSIPVToInTree(pv *v1.Persisten
 
-	// TODO: Take the zone/regional information and stick it into the label.
+	// translate CSI topology to In-tree topology for rollback compatibility
+	if err := translateTopologyFromCSIToInTree(pv, GCEPDTopologyKey); err != nil {
```
we may want to add the labels too in case people have tooling on the kubernetes labels.

Jiawei0227 replied:
The labels of migrated PV already exist in the current case even before it goes to label admission controller.

msau42 replied:
The PV.Labels are set?

Jiawei0227 replied:
Oops, it is not... Sorry I had a misunderstanding. Will add it.

Jiawei0227 replied:
I add the zone labels in if it does not exist. But I dont know if we can derive region info from Zone values for all the plugins. So I did not add the region label.

An alternative solution is to add a `generateRegionFromZone` func as a parameter to let each plugin pass a region parser.

msau42 replied:
We only have 5 plugins to support. I think it's not that critical to have a common util function and each plugin can parse region their own ways.

msau42 commented on staging/src/k8s.io/csi-translation-lib/plugins/gce_pd.go:
```diff
@@ -286,7 +291,10 @@ func (g *gcePersistentDiskCSITranslator) TranslateCSIPVToInTree(pv *v1.Persisten
+	}
 
 	pv.Spec.CSI = nil
 	pv.Spec.GCEPersistentDisk = gceSource
```
I think we also have to enhance the storageclass translation to handle AllowedTopologies having beta/ga labels

msau42 replied:
nm i saw this was done in the common library

msau42 commented on staging/src/k8s.io/csi-translation-lib/plugins/in_tree_volume.go:
```diff
@@ -165,6 +219,57 @@ func translateTopology(pv *v1.PersistentVolume, topologyKey string) error {
+
+	csiTopologyZoneValues := getTopologyValues(pv, topologyKey)
+
+	// Frist remove the CSI topology Key
```
instead of remove/add, can we use replaceTopology?

Jiawei0227 replied:
This function assumes that we always have one CSI Topology label in the NodeAffinity. But we need to supply both zone/region labels. So I think a remove and adding new labels seems more clear what we are trying to achieve.

wongma7 commented on staging/src/k8s.io/csi-translation-lib/plugins/gce_pd.go:
```diff
@@ -215,7 +215,12 @@ func (g *gcePersistentDiskCSITranslator) TranslateInTreePVToCSI(pv *v1.Persisten
 		return nil, fmt.Errorf("pv is nil or GCE Persistent Disk source not defined on pv")
 	}
 
+	// depend on which version it migrates from, the label could be failuredomain beta or topology GA version
```
is it possible for user to manually set the GA label even if provisioner doesn't? if so, this logic could be useful for all.

I'm double checking if ebs provisioned PV would need it, seems like ebs and pd in-tree provisioner set beta label.

gce: https://github.com/kubernetes/kubernetes/blob/819ff9b0870506b889e156adc5e66e314f8d148e/staging/src/k8s.io/legacy-cloud-providers/gce/gce_disks.go#L850 

aws: https://github.com/kubernetes/kubernetes/blob/48a0ef6a396f84ec93a8e68d9843c6fe27c4e848/staging/src/k8s.io/legacy-cloud-providers/aws/aws.go#L2726


(if ebs needs it too, i can handle in own pr)

Jiawei0227 replied:
This logic that we are commenting here is a special logic for gce-pd to generate vol-id. 
For general label and topology handling, we have logic in the `translateTopologyFromInTreeToCSI` which will handle the label translation. 

For gce-pd specific topology label migration I will open a new PR once this is merged. And I will fix the gce-pd label in that PR to migrate to GA. I believe ebs will also have a separate PR to do the same.

wongma7 replied:
got it, just af ew lines below i see the generate code for volid. and ebs won't have this since zone in volId URI is optional...thanks for explaining

wongma7 commented on staging/src/k8s.io/csi-translation-lib/plugins/in_tree_volume.go:
```diff
@@ -124,38 +124,85 @@ func addTopology(pv *v1.PersistentVolume, topologyKey string, zones []string) er
+
+// translateTopologyFromInTreeToCSI converts existing zone labels or in-tree topology to CSI topology.
 // In-tree topology has precedence over zone labels.
-func translateTopology(pv *v1.PersistentVolume, topologyKey string) error {
```
btw do you know come this function only translated zone and not region? it looks like azuredd and cinder provisioners do set region, so the info gets lost on translation, potentially it is a bug for them.

OTOH, your new translateTopologyFromCSIToInTree does handle region makes more sense to me.

Jiawei0227 replied:
Hmm, I am not familiar with those two so I could not tell if this is going to be a problem. But I guess their driver has its own CSI migration test and if those tests are working fine. Then it should be good?

wongma7 replied:
Yeah let's leave it as is, I may open a separate issue if necessary but I was just curious since now there is some asymmetry between translateTopologyFromCSIToInTree and translateTopologyFromInTreeToCSI.

msau42 replied:
cc @andyzhangx @gnufied for region questions on azure and openstack.

At least in aws and gcp cases, you can derive the region from the zone.

msau42 commented on staging/src/k8s.io/csi-translation-lib/plugins/in_tree_volume.go:
```diff
@@ -165,6 +212,101 @@ func translateTopology(pv *v1.PersistentVolume, topologyKey string) error {
+// isTopologyLabelGA checks if the topology label used in this PV is GA
+// and return the zone/region label used. If no labels are found, check
+// NodeAffinity to find the label
+func isTopologyLabelGA(pv *v1.PersistentVolume) (bool, string, string) {
```
a better name for this function may be something like `getTopologyLabel`. Does it need to return a bool? It always matches with the string.

Jiawei0227 replied:
The bool might become handy in some cases. It does not hurt to return it here I think.

msau42 commented on staging/src/k8s.io/csi-translation-lib/plugins/in_tree_volume.go:
```diff
@@ -165,6 +212,101 @@ func translateTopology(pv *v1.PersistentVolume, topologyKey string) error {
+	_, zoneGA := pv.Labels[v1.LabelTopologyZone]
+	_, regionGA := pv.Labels[v1.LabelTopologyRegion]
+	if zoneGA || regionGA {
+		return true, v1.LabelTopologyZone, v1.LabelTopologyRegion
```
Is it possible the label could be GA, but the NodeAffinity could be beta? (or vice versa) It may be safer to look for the NodeAffinity label first?

Jiawei0227 replied:
Yeah... I thought about it and it can become messy. For example, what if there is a skew like this. I think you are right, we should check for NodeAffinity first to make sure.

Jiawei0227 replied:
Done

msau42 commented on staging/src/k8s.io/csi-translation-lib/plugins/in_tree_volume.go:
```diff
@@ -165,6 +212,101 @@ func translateTopology(pv *v1.PersistentVolume, topologyKey string) error {
+		if regionOK {
+			err := addTopology(pv, regionLabel, []string{regionLabelVal})
+			if err != nil {
+				return fmt.Errorf("Failed to add Kubernetes topology zone to PV NodeAffinity. %v", err)
```
update error message for region

Jiawei0227 replied:
Done

msau42 commented on staging/src/k8s.io/csi-translation-lib/plugins/in_tree_volume.go:
```diff
@@ -165,6 +212,121 @@ func translateTopology(pv *v1.PersistentVolume, topologyKey string) error {
+	if !zoneOK {
+		pv.Labels[zoneLabel] = csiTopologyZoneValStr
+	}
+	// Unless we are sure we can derive the region info from zone info. Let's leave the region info for now.
```
Since region is more complicated to handle, what if we rename this function to handle only zone, and leave it up to each plugin to handle region as they wish?

Jiawei0227 replied:
oops... sorry I did not see this, I have already adopt the function in a way to accept a regionParser function to handle region. And if nil is passing, it will skip the region part. Please let me know if that makes sense or not.

msau42 commented on staging/src/k8s.io/csi-translation-lib/plugins/in_tree_volume.go:
```diff
@@ -165,6 +212,121 @@ func translateTopology(pv *v1.PersistentVolume, topologyKey string) error {
+				}
+			}
+		} else {
+			// nothing is in the label, we try our best to add the CSI topology zone in
```
I think we should prioritize CSI first before the PV label. Someone would have to manually create their own PV or have their own CSI driver to get into the scenario where a CSI PV object would have a intree label on it.

Jiawei0227 replied:
See comments above.

msau42 replied:
Sorry which comment are you referring to?

Jiawei0227 replied:
Sorry wrong reply, lol. I thought I was replying to another comment. NVM

So here, you mean if there is a CSI topology and also PV labels. We should use the CSI topology to generate Kubernetes topology first?

msau42 replied:
Yes, we should prefer using what's in CSI topology over PV labels because in the normal case we expect only CSI topology to exist. If PV label exists, something unusual is going on.

Jiawei0227 replied:
Done.

msau42 commented on staging/src/k8s.io/csi-translation-lib/plugins/in_tree_volume.go:
```diff
@@ -302,27 +305,56 @@ func translateTopologyFromCSIToInTree(pv *v1.PersistentVolume, topologyKey strin
+			pv.Labels[zoneLabel] = csiTopologyZoneValStr
+		}
+		if !regionOK && regionParser != nil {
+			regionVal, err := regionParser(csiTopologyZoneValues)
```
It may not always be the case that you can derive region from zone value. It may be better to just pass the whole original PV object in, and potentially let the plugin extract region from their own topology label.

Jiawei0227 replied:
Done.

Jiawei0227 commented:
/retest

Jiawei0227 commented on staging/src/k8s.io/csi-translation-lib/plugins/in_tree_volume.go:
```diff
@@ -88,23 +88,29 @@ func replaceTopology(pv *v1.PersistentVolume, oldKey, newKey string) error {
 			if r.Key == key {
-				values = append(values, r.Values...)
+				for _, v := range r.Values {
+					values[v] = true
```
Because we will get duplicate values. For example, a PV has two NodeSelectorTerms and each of them has a region kubernetes topology which points to the same region.

Jiawei0227 commented on staging/src/k8s.io/csi-translation-lib/plugins/in_tree_volume.go:
```diff
@@ -165,6 +218,153 @@ func translateTopology(pv *v1.PersistentVolume, topologyKey string) error {
+	// Check the NodeAffinity first for the topology version
+	zoneGA := TopologyKeyExist(v1.LabelTopologyZone, pv.Spec.NodeAffinity)
+	regionGA := TopologyKeyExist(v1.LabelTopologyRegion, pv.Spec.NodeAffinity)
+	if zoneGA || regionGA {
```
Yeah, firstly it will be very unexpected... But it could happen, my criteria here is that if it has a GA version topology, then it should be treated as GA supported.

msau42 replied:
I guess what if zone label is beta, but region label is GA. Then we will fail to fine the zone because we're looking for GA zone label?

Jiawei0227 replied:
Hmm, so do you suggest we return zone label version and region label version separately? Like we can return GA version region label and beta version zone label in the case you described?

msau42 replied:
That's a good idea. I think it will simplify the logic too

Jiawei0227 replied:
As we discussed offline, it would be good to not overcomplicate this, so I will keep the old logic which will assume Zone and Region labels are in the same version.

msau42 replied:
Sorry coming back to this again. Seeing `zone || region` makes me asks questions like "what if zone was false?". Since we care about zone most importantly, what if we only check zone, and then return the corresponding region version based off of what we found? (ie, if we found beta zone, then return beta region without checking)

Jiawei0227 replied:
Done

Jiawei0227 commented on staging/src/k8s.io/csi-translation-lib/plugins/in_tree_volume.go:
```diff
@@ -124,38 +130,85 @@ func addTopology(pv *v1.PersistentVolume, topologyKey string, zones []string) er
+		for i, nodeSelectorRequirement := range nsrequirements {
+			if nodeSelectorRequirement.Key == topologyKey {
+				index = i
+				break
```
okay, I changed the logic to remove the key from:
1. Multiple terms(This will be the regional PD case I think)
2. Single term with duplicate key match expression(although I doubt this will happen)

Also add removeTopology test

Jiawei0227 committed:
A complete new way
- Instead of using a remove/add way to deal with the
topology label, go directly with the replace topology
way so that we dont have to care all the corner case.
Also, since the translateTopologyFromCSIToInTree only
get called from csi-provisioner, there should not be
very odd PV object spec.

- Change the beheavior translateTopologyFromInTreeToCSI
so that it will replace zone label to CSI Topology label
and also replace any beta region label to GA label

Jiawei0227 commented:
@msau42 I thought about this more offline. And I thought you are right, it would be easier if we use replaceTopology instead of remove/add. I changed all the corresponding part and write completely new function in the new commit. To summarize the behavior in `translateTopologyFromCSIToInTree`:

1. We will replace all the CSITopology to Kubernetes Zone Topology(And combine existing Kubernetes Zone Topology if any exists already)
2. We will add region topology if a `regionTopologyHandler` function is passed by the plugin. So each plugin can handle their region logic separately. Specifically for gce-pd, I prepared a gceRegionTopologyHandler which basically goes over each NodeAffinity term, and generate region val for each term. So this also handles the case where we have multiple region in different terms.
3. Add zone and region labels to the PV if it does not exist.

Please let me know if this model is easier to understand and follow.

Jiawei0227 commented on staging/src/k8s.io/csi-translation-lib/plugins/in_tree_volume.go:
```diff
@@ -165,6 +229,126 @@ func translateTopology(pv *v1.PersistentVolume, topologyKey string) error {
+	if regionTopologyHandler != nil {
+		err = regionTopologyHandler(pv)
+		if err != nil {
+			// let's make less strict on this one. Even if there is an error in the region processing, just ignore it
```
One thing I want to call out in this commit is that I silently ignore any error from the  regionTopologyHandler just want to make the function less error-prone

andyzhangx commented on plugin/pkg/admission/storage/persistentvolume/label/admission.go:
```diff
@@ -172,15 +172,31 @@ func (l *persistentVolumeLabel) findVolumeLabels(volume *api.PersistentVolume) (
+	domain, domainOK := existingLabels[v1.LabelTopologyZone]
+	region, regionOK := existingLabels[v1.LabelTopologyRegion]
+	// If they dont have GA labels we should check for failuredomain beta labels
+	// TODO: remove this once all the cloud provider change to GA topology labels
```
> remove this once all the cloud provider change to GA topology labels

When will cloud provider change to GA topology labels, for in-tree driver, it's already in maintenance mode, so it would always use beta label? While for CSI driver, it won't use these labels.

Jiawei0227 replied:
We will move in-tree to GA label in this quarter as CSI migration is likely slip and the Beta label will probably be removed this or next release. I will make a PR about making gce-pd label to GA after this PR merged.

andyzhangx commented:
/test pull-kubernetes-e2e-azure-disk-vmss
/test pull-kubernetes-e2e-azure-disk

verult commented on staging/src/k8s.io/csi-translation-lib/plugins/in_tree_volume.go:
```diff
@@ -76,38 +77,85 @@ const (
+		return nil
+	}
 	for i := range pv.Spec.NodeAffinity.Required.NodeSelectorTerms {
+		newKeyValsMap := make(map[string]bool)
```
nit: `newKeyValsMap` -> `newKeyValsSet`

Jiawei0227 replied:
Done

verult commented on staging/src/k8s.io/csi-translation-lib/plugins/in_tree_volume.go:
```diff
@@ -76,38 +77,85 @@ const (
+
+		// Combine any duplicated keys
+		if len(requirementsToRemove) > 1 {
+			// This means there are duplicated new Labels, we shall combine them together
```
If only a single entry of the new key exists in NodeAffinity, based on the logic above, it would be appended to `requirementsToRemove` even though it's not a duplicate right?

Jiawei0227 replied:
It will be added but the `if len(requirementsToRemove) > 1` logic here makes sure that we do not do unnecessary removal.

verult commented on staging/src/k8s.io/csi-translation-lib/plugins/in_tree_volume.go:
```diff
@@ -76,38 +77,85 @@ const (
+			}
+			// after remove all, append the new value
+			var newKeyVals []string
+			for k := range newKeyValsMap {
```
nit: `k` -> `val`, since `k` could be confused with topology key

Jiawei0227 replied:
Done

verult commented on staging/src/k8s.io/csi-translation-lib/plugins/in_tree_volume.go:
```diff
@@ -76,38 +77,85 @@ const (
 			if r.Key == oldKey {
 				pv.Spec.NodeAffinity.Required.NodeSelectorTerms[i].MatchExpressions[j].Key = newKey
 			}
+			if r.Key == newKey || r.Key == oldKey {
```
Does this approach work for deduplication: create a new empty list for `MatchExpressions`, and only append values that don't satisfy this if statement

Jiawei0227 replied:
Sorry I might not get the case you mentioned here. Can you give a small example? The logic here is that for each term, we iterate the MatchExpression and see if there is a newKey/oldKey exist, if it does, we add it to the `requirementsToRemove` candidate. And after iteration, we check the size of `requirementsToRemove`. It it is greater than 1. It means we have duplication.

If we have a MatchExpressions that has no newKey/oldKey, then it will just skip this term and keep working on next term.

Jiawei0227 replied:
Since we are not remove duplication here. I am just going to keep the original logic which is to replace the old key with new key.

verult commented on staging/src/k8s.io/csi-translation-lib/plugins/in_tree_volume.go:
```diff
@@ -124,38 +172,54 @@ func addTopology(pv *v1.PersistentVolume, topologyKey string, zones []string) er
 	}
 
-	zones := getTopologyZones(pv, v1.LabelFailureDomainBetaZone)
+	_, zoneLabel, _ := getTopologyLabel(pv)
```
Should region keys be removed from CSI NodeAffinity?

Jiawei0227 replied:
Good question. I did not change much on the `translateTopologyFromInTreeToCSI` case. So the original behavior is to keep the region keys. @msau42 What do you think. Should we remove it when in-tree -> CSI? Well, actually in the current implementation, the kubernetes zone topology labels is not guaranteed to be removed because what this function does is just to make sure the CSITopologyKey exists, and it dont really care about Kubernetes topology.

msau42 replied:
When is intree -> csi called, and when would it matter? My initial hunch is the scheduler. And perhaps that might cause issues because a PV could have beta region label on it, which would fail scheduling.

verult replied:
Maybe we could have a region handler here as well, in case a storage provider has special region support?

Jiawei0227 replied:
I think msau mentioned a good point. Suppose that we are migrating an old PV to a new cluster where nodes only have GA region labels. And that might fail the scheduling. So I think the right thing to do here is to remove the beta region label at least/replace the beta region label to GA label as we are sure all the node should have GA label at least.

verult reviewed (changes requested):
Nits are mostly style comments to improve readability - please feel free to ignore where you think is appropriate

Jiawei0227 commented on staging/src/k8s.io/csi-translation-lib/plugins/in_tree_volume.go:
```diff
@@ -124,38 +172,54 @@ func addTopology(pv *v1.PersistentVolume, topologyKey string, zones []string) er
 
-// translateTopology converts existing zone labels or in-tree topology to CSI topology.
+// translateTopologyFromInTreeToCSI converts existing zone labels or in-tree topology to CSI topology.
 // In-tree topology has precedence over zone labels.
```
Done

Jiawei0227 commented on staging/src/k8s.io/csi-translation-lib/plugins/in_tree_volume.go:
```diff
@@ -165,6 +229,126 @@ func translateTopology(pv *v1.PersistentVolume, topologyKey string) error {
 	return nil
 }
 
+// getTopologyLabel checks if the kubernetes topology used in this PV is GA
```
Done

Jiawei0227 commented on staging/src/k8s.io/csi-translation-lib/plugins/in_tree_volume.go:
```diff
@@ -165,6 +229,126 @@ func translateTopology(pv *v1.PersistentVolume, topologyKey string) error {
+// It first check the NodeAffinity to find topology. If nothing is found,
+// It checks the PV labels. If it is empty in both places, it will return
+// GA label by default
+func getTopologyLabel(pv *v1.PersistentVolume) (bool, string, string) {
```
No. I might just remove it. I first made it and thought it might get handy somehow. But since it is not used anywhere. Guess we dont need this extra bool

Jiawei0227 commented on staging/src/k8s.io/csi-translation-lib/plugins/in_tree_volume.go:
```diff
@@ -124,38 +172,54 @@ func addTopology(pv *v1.PersistentVolume, topologyKey string, zones []string) er
+// translateTopologyFromInTreeToCSI converts existing zone labels or in-tree topology to CSI topology.
 // In-tree topology has precedence over zone labels.
-func translateTopology(pv *v1.PersistentVolume, topologyKey string) error {
+func translateTopologyFromInTreeToCSI(pv *v1.PersistentVolume, topologyKey string) error {
```
Done

Jiawei0227 commented on staging/src/k8s.io/csi-translation-lib/plugins/in_tree_volume.go:
```diff
@@ -76,38 +77,85 @@ const (
 )
 
-// replaceTopology overwrites an existing topology key by a new one.
+// replaceTopology overwrites an existing key by a new one.
```
Done

Jiawei0227 commented on staging/src/k8s.io/csi-translation-lib/plugins/in_tree_volume.go:
```diff
@@ -165,6 +229,126 @@ func translateTopology(pv *v1.PersistentVolume, topologyKey string) error {
+	// 2. Take care of region topology if a regionTopologyHandler is passed
+	if regionTopologyHandler != nil {
+		// let's make less strict on this one. Even if there is an error in the region processing, just ignore it
+		_ = regionTopologyHandler(pv)
```
Yeah, my thought is to add a warning or error log. But then I found out there is no logging in the csi-translation-lib. So I am also not too sure if we can just abort the error. @msau42 Could you suggest what we do here? Should I add logging? Or just silently ignore any error and do not return an error instead?

msau42 replied:
I think returning error is fine

Jiawei0227 replied:
Okay, I will return error if the regionHandler fail then

Jiawei0227 commented on staging/src/k8s.io/csi-translation-lib/plugins/in_tree_volume.go:
```diff
@@ -76,38 +77,85 @@ const (
 
-// getTopologyZones returns all topology zones with the given key found in the PV.
-func getTopologyZones(pv *v1.PersistentVolume, key string) []string {
+// getTopologyValues returns all unique topology values with the given key found in the PV.
```
Done

Jiawei0227 commented on staging/src/k8s.io/csi-translation-lib/plugins/in_tree_volume.go:
```diff
@@ -165,6 +229,126 @@ func translateTopology(pv *v1.PersistentVolume, topologyKey string) error {
+//    combine existing kubernetes zone together
+// 2. Process and generate region topology if a regionTopologyHandler is passed
+// 3. Add Kubernetes Topology labels(zone and region) if they do not exist
+func translateTopologyFromCSIToInTree(pv *v1.PersistentVolume, topologyKey string, regionTopologyHandler regionTopologyHandler) error {
```
Done

Jiawei0227 commented on staging/src/k8s.io/csi-translation-lib/plugins/in_tree_volume.go:
```diff
@@ -165,6 +229,126 @@ func translateTopology(pv *v1.PersistentVolume, topologyKey string) error {
+		}
+		_, zoneOK := pv.Labels[zoneLabel]
+		if !zoneOK {
+			zoneValStr := strings.Join(zoneVals, labelMultiZoneDelimiter)
```
zoneVals have already been sorted in the `getTopologyValues` return. I will add a comment in the func

Jiawei0227 commented on staging/src/k8s.io/csi-translation-lib/plugins/in_tree_volume.go:
```diff
@@ -165,6 +229,126 @@ func translateTopology(pv *v1.PersistentVolume, topologyKey string) error {
+		}
+		_, regionOK := pv.Labels[regionLabel]
+		if !regionOK {
+			regionValStr := strings.Join(regionVals, labelMultiZoneDelimiter)
```
Good question. This is just in case some crazy driver provides ability like multiple terms:
1. region A, zone 1
2. region B, zone 2

If it only return one regionVals it will do no harm to GCP so I think it should be fine.

msau42 replied:
I think adding region labels should be the responsibility of the region handler. I'm not sure if any volume type we're dealing with supports multi region.

Jiawei0227 replied:
Okay, I will move this logic to gce then.

Jiawei0227 commented on staging/src/k8s.io/csi-translation-lib/plugins/in_tree_volume.go:
```diff
@@ -165,6 +229,126 @@ func translateTopology(pv *v1.PersistentVolume, topologyKey string) error {
+
+// translateTopologyFromCSIToInTree translate a CSI topology to
+// Kubernetes topology and add topology labels to it. Note that this function
+// will only work for plugin with single topologyKey. If a plugin has
```
Done

Jiawei0227 commented on staging/src/k8s.io/csi-translation-lib/plugins/gce_pd.go:
```diff
@@ -369,6 +377,51 @@ func pdNameFromVolumeID(id string) (string, error) {
+
+	// process each term
+	for index, nodeSelectorTerm := range pv.Spec.NodeAffinity.Required.NodeSelectorTerms {
+		// In the first loop, see if regionLabel already exist
```
Yeah, that's true. This just keeps the potential if we ever want to do something with the kubernetes topology in csi-provisioner? It does not hurt to check just to make sure we dont append a label that already exist.

Jiawei0227 commented on staging/src/k8s.io/csi-translation-lib/plugins/in_tree_volume.go:
```diff
@@ -76,38 +77,85 @@ const (
-// replaceTopology overwrites an existing topology key by a new one.
+// replaceTopology overwrites an existing key in NodeAffinity by a new one.
+// If there are any newKey already exist in a term, we will combine the
+// replaced key Values with the existing ones
```
Ohh, currently in my implementation I will combine them together... So basically if we have
`gke.topology.io/zone=zone1` and `topology.kubernetes.io/zone=zone2`
I will combine it to `topology.kubernetes.io/zone=zone1 and zone2`

msau42 replied:
hmmmm I think that is changing the semantics of the NodeAffinity.

If they're combined in the same expression, then it is ORed together, which means the volume must be in "zone1 OR zone2".

If they're in different expressions, we AND them together, which means the volume must be in "zone1 AND zone2". Which at least with the volumes we're dealing with, would not make sense.

msau42 replied:
Thinking about this more, maybe we should not try to reduce terms here. It would be very strange (and probably non functional) if a user created a PV like this. Instead of trying to fix their problem, we can just leave it as is and let them fix it. And in the case of where it's the same values, having 2 ANDs of the same thing should be fine (we can also try it out to make sure the scheduler handles it properly)

Jiawei0227 replied:
> If they're combined in the same expression, then it is ORed together, which means the volume must be in "zone1 OR zone2". If they're in different expressions, we AND them together, which means the volume must be in "zone1 AND zone2". 

Yeah, this makes sense to me. I did not think of that and that would definitely be a weird case. Hmm, not remove duplication is definitely not so UX friendly. But think about this more, it only get called in the csi-provisioner and right now we do not add the Kubernetes topology there so maybe we are over thinking. I will just remove the whole logic then. Sounds like we just need to replace it.

Jiawei0227 replied:
Done

Jiawei0227 commented on staging/src/k8s.io/csi-translation-lib/plugins/in_tree_volume.go:
```diff
@@ -124,41 +172,179 @@ func addTopology(pv *v1.PersistentVolume, topologyKey string, zones []string) er
-	)
+	// add the CSI topology to each term
+	for i := range pv.Spec.NodeAffinity.Required.NodeSelectorTerms {
+		pv.Spec.NodeAffinity.Required.NodeSelectorTerms[i].MatchExpressions = append(
```
I thought array does not need initialization? https://play.golang.org/p/r7D8bx7DSXc

Jiawei0227 commented on staging/src/k8s.io/csi-translation-lib/plugins/in_tree_volume.go:
```diff
@@ -76,38 +77,85 @@ const (
 		}
 	}
-	return values
+	var re []string
```
Hmm, because values is a map so that we can get rid of duplicate values.
There will be cases in different terms we have same key, value so there will be duplication. Sort is because we will probably use the slice to generate PV label and we want that in order.

msau42 replied:
Ok, that seems reasonable. Can you add some comments explaining those?

Also it may be worth commenting that this function collapses multiple zones into a list that is ORed. It assumes the plugin doesn't support a constraint like "zone1" and "zone2"

Jiawei0227 replied:
Done

Jiawei0227 commented on staging/src/k8s.io/csi-translation-lib/plugins/gce_pd.go:
```diff
@@ -369,6 +377,67 @@ func pdNameFromVolumeID(id string) (string, error) {
+			continue
+		}
+		// If no regionLabel found, generate region label from the zoneLabel we collect from this term
+		regionVal, err := getRegionFromZones(zoneVals)
```
I actually think a lot of other cloud-provider like aws/azure an reuse this function if we allow getRegionFromZones function to be a parameter. The other logic should be the same for other cloud-providers. But I guess it is fine we keep it this way. We can let them to decide if they want to use this or not.

msau42 replied:
That sounds like a good idea.

msau42 commented on staging/src/k8s.io/csi-translation-lib/plugins/in_tree_volume.go:
```diff
@@ -124,41 +146,171 @@ func addTopology(pv *v1.PersistentVolume, topologyKey string, zones []string) er
+	if zoneBeta := TopologyKeyExist(v1.LabelFailureDomainBetaZone, pv.Spec.NodeAffinity); zoneBeta {
+		return v1.LabelFailureDomainBetaZone, v1.LabelFailureDomainBetaRegion
+	}
+	if regionGA := TopologyKeyExist(v1.LabelTopologyRegion, pv.Spec.NodeAffinity); regionGA {
```
I was thinking we don't need to check for region at all. Because if zone is missing, then we're in trouble.

Jiawei0227 replied:
Sounds good. I can remove the region related checks and only check zone

msau42 commented on staging/src/k8s.io/csi-translation-lib/plugins/in_tree_volume.go:
```diff
@@ -124,41 +146,171 @@ func addTopology(pv *v1.PersistentVolume, topologyKey string, zones []string) er
+// regionTopologyHandler is a function to add region topology NodeAffinity
+// and labels for the given PV. It assumes the Zone NodeAffinity already exists
+// If regionTopologyHandler is nil, no region NodeAffinity will be added
+// If regionTopologyHandler returns err, it will be ignored.
```
comment needs updating

Jiawei0227 replied:
Done

andyzhangx commented:
/test pull-kubernetes-e2e-azure-disk-vmss
/test pull-kubernetes-e2e-azure-disk

msau42 commented:
/lgtm
/approve

Thanks!

msau42 commented:
/assign @andrewsykim

Jiawei0227 commented:
I rebased commits to make it cleaner. No new changes are introduced.

andyzhangx commented:
/test pull-kubernetes-e2e-azure-disk-vmss
/test pull-kubernetes-e2e-azure-disk

msau42 commented:
/lgtm

Jiawei0227 commented:
@andrewsykim Gentle ping. Hey Andrew, can you help to take a look? This is blocking the following steps of updating the topology label to GA. Thanks a lot

andrewsykim reviewed (commented):
/approve 

for k8s.io/cloud-provider, thanks!
